test/multipart/test_encoder.py

```diff
@@ -0,0 +1,287 @@
+
+    def test_streams_its_data(self) -> None:
+        large_file = LargeFileMock()
+        parts: typing.Mapping[str, str | typing.BinaryIO] = {
+            "some field": "value",
+            "some file": large_file,
+        }
+        encoder = MultipartEncoder(parts)
+        total_size = len(encoder)
+        read_size = 1024 * 1024 * 128
+        already_read = 0
+        while True:
+            read = encoder.read(read_size)
+            already_read += len(read)
+            if not read:
+                break
+
+        assert encoder._buffer.tell() <= read_size
+        assert already_read == total_size
+
+    def test_length_is_correct(self) -> None:
+        encoded = filepost.encode_multipart_formdata(self.parts, self.boundary)[0]
+        assert len(encoded) == len(self.instance)
+
+    def test_encodes_with_readable_data(self) -> None:
+        s = io.BytesIO(b"value")
+        m = MultipartEncoder([("field", s)], boundary=self.boundary)
+        assert m.read() == (
+            b"--this-is-a-boundary\r\n"
+            b'Content-Disposition: form-data; name="field"\r\n\r\n'
+            b"value\r\n"
+            b"--this-is-a-boundary--\r\n"
+        )
+
+    def test_reads_open_file_objects(self) -> None:
+        with open(__file__, "rb") as fd:
+            m = MultipartEncoder([("field", "foo"), ("file", fd)])
+            assert m.read() is not None
+
+    def test_reads_open_file_objects_with_a_specified_filename(self) -> None:
+        with open(__file__, "rb") as fd:
+            m = MultipartEncoder(
+                [("field", "foo"), ("file", ("filename", fd, "text/plain"))]
+            )
+            assert m.read() is not None
+
+    def test_handles_encoded_unicode_strings(self) -> None:
+        m = MultipartEncoder(
+            [
+                (
+                    "field",
+                    b"this is a unicode string: \xc3\xa9 \xc3\xa1 \xc7\xab \xc3\xb3",
+                )
+            ]
+        )
+        assert m.read() is not None
+
+    def test_handles_uncode_strings(self) -> None:
+        s = b"this is a unicode string: \xc3\xa9 \xc3\xa1 \xc7\xab \xc3\xb3"
+        m = MultipartEncoder([("field", s.decode("utf-8"))])
+        assert m.read() is not None
+
+    def test_regression_1(self) -> None:
+        """Ensure issue #31 doesn't ever happen again."""
+        fields: dict[str, str | tuple[str, typing.BinaryIO]] = {"test": "t" * 100}
+
+        for x in range(30):
+            fields["f%d" % x] = ("test", open(__file__, "rb"))
+
+        m = MultipartEncoder(fields=fields)
+        total_size = len(m)
+
+        blocksize = 8192
+        read_so_far = 0
+
+        while True:
+            data = m.read(blocksize)
+            if not data:
+                break
+            read_so_far += len(data)
+
+        assert read_so_far == total_size
+
+        for k, v in fields.items():
+            if k != "test" and hasattr(v[1], "close"):
+                v[1].close()
```

Comment:
These lines are new – pytest was raising an error about unclosed files, from the `open()` on L175.

Reply:
The alternative is doing this with a `pytest.mark.filterwarnings` decorator, which would be a little clearer – I might change this.

Reply:
Done!